R/return_linter.R

```diff
@@ -0,0 +1,177 @@
+#' Return linter
+#'
+#' This linter checks functions' [return()] expressions.
+#'
+#' @param return_style Character string naming the return style. `"implicit"`,
+#'   the default, enforeces the Tidyverse guide recommendation to leave terminal
+#'   returns implicit. `"explicit"` style requires that `return()` always be
+#'   explicitly supplied.
+#' @param return_functions Character vector of functions that are accepted as terminal calls
+#'   when `return_style = "explicit"`. These are in addition to exit functions
+#'   from base that are always allowed: [stop()], [q()], [quit()], [invokeRestart()],
+#'   `tryInvokeRestart()`, [UseMethod()], [NextMethod()], [standardGeneric()],
+#'   [callNextMethod()], [.C()], [.Call()], [.External()], and [.Fortran()].
+#' @param except Character vector of functions that are not checked when
+#'   `return_style = "explicit"`. These are in addition to namespace hook functions
+#'   that are never checked: `.onLoad()`, `.onUnload()`, `.onAttach()`, `.onDetach()`,
+#'   and `.Last.lib()`.
+#'
+#' @examples
+#' # will produce lints
+#' code <- "function(x) {\n  return(x + 1)\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = return_linter()
+#' )
+#'
+#' code <- "function(x) {\n  x + 1\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = return_linter(return_style = "explicit")
+#' )
+#'
+#' # okay
+#' code <- "function(x) {\n  x + 1\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = return_linter()
+#' )
+#'
+#' code <- "function(x) {\n  return(x + 1)\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = return_linter(return_style = "explicit")
+#' )
+#'
+#'
+#' @evalRd rd_tags("return_linter")
+#' @seealso
+#'  - [linters] for a complete list of linters available in lintr.
+#'  - <https://style.tidyverse.org/functions.html?q=return#return>
+#' @export
+return_linter <- function(
+    return_style = c("implicit", "explicit"),
+    return_functions = NULL,
+    except = NULL) {
+  return_style <- match.arg(return_style)
+
+  if (return_style == "implicit") {
+    xpath <- "
+      (//FUNCTION | //OP-LAMBDA)
+      /following-sibling::expr[1][*[1][self::OP-LEFT-BRACE]]
+      /expr[last()][
+        expr[1][
+          not(OP-DOLLAR or OP-AT)
+          and SYMBOL_FUNCTION_CALL[text() = 'return']
+        ]
+      ]
+    "
+    msg <- "Use implicit return behavior; explicit return() is not needed."
+  } else {
+    # See `?.onAttach`; these functions are all exclusively used for their
+    #   side-effects, so implicit return is generally acceptable
+    default_except <- c(".onLoad", ".onUnload", ".onAttach", ".onDetach", ".Last.lib")
```

Comment:
just noticed we could probably re-use this:

https://github.com/r-lib/lintr/blob/1c36e0d03b3f39a13f1a79d173548c2f75ec9bf9/R/object_name_linter.R#L186-L194